src/protocol/fluvio-protocol-codec/src/codec.rs

```diff
@@ -101,110 +102,146 @@ mod test {
-            let (_, mut stream) = framed.split();
-            for _ in 0..3u16 {
-                if let Some(value) = stream.next().await {
-                    debug!("client :received first value from server");
-                    let mut bytes = value.expect("bytes");
-                    debug!("client: received bytes len: {}", bytes.len());
-                    assert_eq!(bytes.len(), 9, "total bytes is 6");
-                    let mut decoded_values: Vec<u8> = vec![];
-                    decoded_values
-                        .decode(&mut bytes, 0)
-                        .expect("vector decoding failed");
-                    assert_eq!(decoded_values.len(), 5);
-                    assert_eq!(decoded_values[0], 1);
-                    assert_eq!(decoded_values[1], 2);
-                    debug!("all test pass");
-                } else {
-                    panic!("no first value received");
-                }
+            let (mut sink, _) = framed.split();
+
+            // send 2 times in order
+            for _ in 0..2u16 {
+                //  debug!("server encoding original vector with len: {}", data.len());
+                let mut buf = vec![];
+                data.encode(&mut buf, 0)?;
+                debug!(
+                    "server: writing to client vector encoded len: {}",
+                    buf.len()
+                );
+                assert_eq!(buf.len(), 9); //  4(array len)+ 5 bytes
+
+                // write buffer length since when the encoder is working
+                // over raw data, the length has to be send.
+                let mut len_buf = vec![];
+                let len = buf.len() as i32;
+                len.encode(&mut len_buf, 0).expect("encoding");
+                sink.send(Bytes::from(len_buf)).await.expect("sending");
+
+                sink.send(Bytes::from(buf)).await.expect("sending");
+            }
+
+            // last one, we send split send, to test partial send
+            {
+                let mut buf = vec![];
+                data.encode(&mut buf, 0)?;
+                debug!(
+                    "server: writing to client vector encoded len: {}",
+                    buf.len()
+                );
+                assert_eq!(buf.len(), 9); //  4(array len)+ 5 bytes
+
+                // write buffer length since when the encoder is working
+                // over raw data, the length has to be send.
+                let mut len_buf = vec![];
+                let len = buf.len() as i32;
+                len.encode(&mut len_buf, 0).expect("encoding");
+                sink.send(Bytes::from(len_buf)).await.expect("sending");
+
+                // split buf into two segments, decode should reassembly them
+                let buf2 = buf.split_off(5);
+                sink.send(Bytes::from(buf)).await.expect("sending");
+                fluvio_future::timer::sleep(time::Duration::from_millis(10)).await;
```

Comment:
this is not quite right.  this should test partial codec delivery.  so can't user sink here but use tcp_stream directly